docs: add the Python Repo section into OEP-58 docs (#366)
openedx/docs.openedx.org: source/developers/how-tos/enable-translations-new-repo.rst

Enabling Translations on a New Repo
###################################

.. contents::
 :local:
 :depth: 2

Quickstart

> *********

Author: Done @feanil!

3. Ensure ``make extract_translations`` produces ``django.po`` (in addition to ``djangojs.po`` if applicable).

4. Add en entry of the repository in the `extract-translation-source-files.yml <https://github.com/openedx/openedx-translations/blob/2566e0c9a30d033e5dd8d05d4c12601c8e37b4ef/.github/workflows/extract-translation-source-files.yml#L36-L43>`_ openedx-translations GitHub workflow file.

  Add a new entry under the ``python-translations`` section. For example for the XBlock located in `github.com/openedx/MyAIBotXBlock <https://github.com/openedx/MyXBlock>`_ and has the XBlock Python module name ``my_ai_bot_xblock`` it should have the following entry::

    python-translations:
      strategy:
        matrix:
          repo:
            ...
            - repo_name: MyAIBotXBlock
              python_module_name: my_ai_bot_xblock
            ...

  When adding a theme or any other module that's not an Open edX plugin or an XBlock the entry should
  added without ``python_module_name`` as demonstrated below::

    python-translations:
      strategy:
        matrix:
          repo:
            ...
            - repo_name: credentials-themes
            ...

>   Create a pull request in the ``openedx-translations`` repo and wait for review and approval.
> 
> 5. Once the pull request you created in the ``openedx-translations`` repo is merged the GitHub/Transifex workflow will
>    automatically perform the following:

Reviewer: I think bringing all the PR stuff together under 5 makes sense, something like:

```suggestion
5.  Create a pull request in the ``openedx-translations`` repo and wait for review and approval. Once the pull request is merged the GitHub/Transifex workflow will
   automatically:
```

>    - Add the extracted source (English) translation from the XBlock GitHub repository to the

Reviewer: ```suggestion
   - Add the extracted source (English) translation files from the repository you added to the
```

     ``openedx-translations`` repo.

>    - Create a new Transifex resource for the XBlock in the ``openedx-translations`` Transifex project.

Reviewer: ```suggestion
   - Create a new Transifex resource for your repo in the ``openedx-translations`` Transifex project.
```

   - Add the translations (Arabic, French, etc) to the GitHub ``openedx-translations`` repo when the Transifex project
     is updated.

>    No additional configuration is needed and the full sync automation log can be found on the following link:

Reviewer: ```suggestion
   No additional configuration is needed. The full sync automation log can be found on the following link:
```

  be used in the repo Makefiles, build and deploy processes to fetch translations from the
  `openedx-translations repo`_.

After configuring the repo as shown in the sections below, the translations workflow will automatically:

- Add the extracted source (English) translation files from the repository you added to the
  `openedx-translations repo`_ in GitHub.
- Create a new Transifex resource for your repo in the `openedx-translations project`_ in Transifex.
- Add the translations (Arabic, French, etc) to the GitHub `openedx-translations repo`_ when translation entries are
  updated by translators.

The full background and rationale for the new workflows and tooling can be found in the
`OEP-58 - Translations Management`_ docs.

The next sections cover the steps to enable translations on a new repo depending its type:

=================================
Python Django Plugins and XBlocks
=================================

This section covers repositories such as XBlocks, Open edX plugins, Python libraries and other Python repositories
except if it's a Django microservice which is covered in the next section.

To include a repo in the translations workflow follow the steps below:

> 1. If you are not working within an existing repository, create a new repo using the: `edx-cookiecutters`_
>    python-library for either ``django-app`` or ``xblock``.
>    If the repo already exists continue to the next steps.

Reviewer: ```suggestion
1. If you are creating a new repository, use the: `edx-cookiecutters`_
   python-library for ``django-app`` or ``xblock``.
```

2. For existing repos which don't have the ``make extract_translations`` command, it can be copied from the
   `edx-cookiecutters`_ Makefile in the corresponding template e.g. ``cookiecutter-xblock`` Makefile for XBlocks.

   **Note:** Some repositories uses ``django-manage makemessages``. The recommendation is to use ``i18n_tool extract``
   because it provides more options and cleans ``.po`` files.
   However, it should be fine to use either as long as the translation files are extracted and working .

3. Run the ``make extract_translations``. Verify that it extracts the files within the ``my_xblock_module/conf/locale``
   directory. The PO file names will vary depending on the use case:

   - The XBlocks will have a ``text.po`` extracted.
   - Django plugins will have ``django.po``.
   - Other plugins may include ``djangojs.po`` if the repo has a ``static`` directory with JavaScript use of
     ``gettext``.

4. Add the repository to `extract-translation-source-files.yml`_ in the `openedx-translations repo`_.

   Add a new entry under the ``python-translations`` section. For example for the XBlock located in

>    ``https://github.com/openedx/MyXBlock`` and has the XBlock Python module name ``my_ai_bot_xblock`` it should have
>    the following entry::
> 
>     python-translations:
>       strategy:
>         matrix:
>           repo:
>             ...
>             - repo_name: MyAIBotXBlock
>               python_module_name: my_ai_bot_xblock
>             ...

Author: ```suggestion
   ``https://github.com/openedx/xblock-drag-and-drop-v2`` and has the XBlock Python 
   module name ``drag_and_drop_v2`` it should have the following entry::

    python-translations:
      strategy:
        matrix:
          repo:
            ...
            - repo_name: xblock-drag-and-drop-v2
              python_module_name: drag_and_drop_v2
            ...
```

   When adding a theme or any other module that's not an Open edX plugin or an XBlock the entry should
   added without ``python_module_name`` as demonstrated below::

    python-translations:
      strategy:
        matrix:
          repo:
            ...
            - repo_name: credentials-themes
            ...

> 5. Create a pull request in the `openedx-translations repo`_ and wait for review and approval.
>    Once the pull request is merged the GitHub/Transifex workflow will be ongoing as described in the
>    Introduction section. No additional configuration is needed.

Reviewer: ```suggestion
5. Create a pull request in the `openedx-translations repo`_
```

> 6. Verify the workflow is syncing the translations properly as described in the Debugging section below.

Reviewer: it'd probably be good to link to the debugging section

> 7. Install the XBlock or plugin in a devstack and run the ``OPENEDX_ATLAS_PULL=true make pull_translations`` in
>    the edx-platform repo to preview the translations in devstack.
> 
>    This step is somewhat oversimplified and assumes you're familiar with the Open edX development workflow whether
>    you're using `Tutor <https://docs.tutor.overhang.io/>`_ or the
>    `Open edX devstack <https://github.com/openedx/devstack/>`_.

Reviewer: ```suggestion
7. Install the XBlock or plugin in your local `Tutor`_ or `devstack`_ environment. Run  ``OPENEDX_ATLAS_PULL=true make pull_translations`` in
   the edx-platform repo to preview the translations.
```

>    `Open edX devstack <https://github.com/openedx/devstack/>`_.

Reviewer: does it make sense to move this step above the PR step? it seems like testing locally is something that should happen before PRing, but the listed command won't work pre-PR

================================
Django Microservice Repos (IDAs)
================================

In the context of Open edX both of the Microservice and Independently Deployable Application (IDA) terms are used
interchangeably. There are many Django microservices in the Open edX ecosystem.
The most common ones are: ``edx-platform`` itself, ``credentials``, ``ecommerce`` and ``course-discovery``.

> 1. If you are not working within an existing repository, create a new repo using the `cookiecutter-django-ida`_
>    template. If the repo already exists continue to the next steps.

Reviewer: ```suggestion
1. If you are creating a new repository, use the `cookiecutter-django-ida`_
   template.
```

TODO: How to enable translations infrastructure on a Django microservice

> 5. Create a pull request in the `openedx-translations repo`_ and wait for review and approval.
>    Once the pull request is merged the GitHub/Transifex workflow will be ongoing as described in the
>    Introduction section. No additional configuration is needed.

Reviewer: ```suggestion
5. Create a pull request in the `openedx-translations repo`_
```

6. Verify the workflow is syncing the translations properly as described in the Debugging section below.

> 7. Run the ``OPENEDX_ATLAS_PULL=true make pull_translations`` to verify that it pulls the translations from the
>    `openedx-translations repo`_ into ``conf/locale`` directory. Depending on the repo
>    ``make static``/``make static.dev`` is usually needed to generate the JavaScript translation files.

Reviewer: ```suggestion
7. Run ``OPENEDX_ATLAS_PULL=true make pull_translations`` to verify translations are pulled from the
   `openedx-translations repo`_ into the ``conf/locale`` directory. To generate JavaScript translation files you will likely also need to run ``make static``/``make static.dev``.
```

>    ``make static``/``make static.dev`` is usually needed to generate the JavaScript translation files.

Reviewer: does it make sense to move this step above the PR step? it seems like testing locally is something that should happen before PRing, but the listed command won't work pre-PR

===========
React Repos
===========

> 1. If you are not working within an existing repository, create a new repo using the `frontend-template-application`_
>    template. If the repo already exists continue to the next steps.

Reviewer: ```suggestion
1. If you are creating a new repository, use `frontend-template-application`_.
```

TODO: How to enable translations infrastructure on a React repo

> 4. Run the ``make extract_translations``. Verify that it extracts the ``src/i18n/transifex_input.json`` file
>    in the micro-frontend repo. This file should be excluded from the repo via the ``.gitignore`` file.

Reviewer: ```suggestion
4. Run ``make extract_translations``. Verify that it creates``src/i18n/transifex_input.json``. This file should be excluded from the repo via the ``.gitignore`` file.
```

> 5. Create a pull request in the `openedx-translations repo`_ and wait for review and approval.
>    Once the pull request is merged the GitHub/Transifex workflow will be ongoing as described in the
>    Introduction section.

Reviewer: ```suggestion
5. Create a pull request in the `openedx-translations repo`_
```

.. contents::
 :local:
 :depth: 2

> ==========

Reviewer 2: Please use the heading structure describe in the Documentation quick-reference instead of using double headings.  https://docs.openedx.org/en/latest/documentors/references/quick_reference_rst.html#headings

management:

- Created the new `openedx-translations project`_ in Transifex.
- Created the new `openedx-translations repo`_ in GitHub.
- Setup two-way links between the Transifex project and the GitHub repo via the
  `Transifex GitHub app <https://github.com/apps/transifex-integration>`_.
- Added new the `atlas <https://github.com/openedx/openedx-atlas>`_ command line tool to pull translations which will
  be used in the repo Makefiles, build and deploy processes to fetch translations from the
  `openedx-translations repo`_.

After configuring the repo as shown in the sections below, the translations workflow will automatically:

- Add the extracted source (English) translation files from the repository you added to the
  `openedx-translations repo`_ in GitHub.
- Create a new Transifex resource for your repo in the `openedx-translations project`_ in Transifex.
- Add the translations (Arabic, French, etc) to the GitHub `openedx-translations repo`_ when translation entries are
  updated by translators.

The full background and rationale for the new workflows and tooling can be found in the
`OEP-58 - Translations Management`_ docs.

> 

Reviewer 2: This section is useful context but I think belongs in a `concepts` doc probably one titled `Translation Technology and Process`.  Then I think you could add a link to that in a `..seealso` section of this document.

Author: I've removed this section and focused on the How-to part.

This section is a summary of the OEP-58 document. So if we're aiming for brevity, a link should suffice.

I have the habit of summerization from the time that I used to write some academic research.

> The next sections cover the steps to enable translations on a new repo depending its type:

Reviewer 2: I don't think you need this line.

=================================
Python Django Plugins and XBlocks
=================================

To include a non-microservice Python repository (such as an XBlock, Open edX plugin, or library) in the translations
workflow:

#. If you are creating a new repository, use the `edx-cookiecutters`_ templates such as ``django-app`` or ``xblock``.

#. For existing repos which don't have the ``make extract_translations`` command, it can be copied from the
   `edx-cookiecutters`_ Makefile in the corresponding template e.g. ``cookiecutter-xblock`` Makefile for XBlocks.

>    **Note:** Some repositories use ``django-manage makemessages``. The recommendation is to use ``i18n_tool extract``

Reviewer 2: Use a `..note` admonition instead: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-note

    python-translations:
      strategy:
        matrix:
          repo:
            ...
            - repo_name: credentials-themes
            ...

#. Create a draft pull request in the `openedx-translations repo`_

#. Follow the instructions in the :ref:`testing-in-your-fork` section to verify new repository configuration.

#. Mark your pull request as ready for review and wait for it to be merged.

#. Verify the workflow is syncing the translations properly as described in the :ref:`debugging-translations` section.

#. Install the XBlock or plugin in your local `Tutor`_ or `devstack`_ environment. Run
   ``OPENEDX_ATLAS_PULL=true make pull_translations`` in the edx-platform repo to preview the translations.

================================
Django Microservice Repos (IDAs)
================================

The terms "Microservice" and "Independently Deployable Application (IDA)" are used interchangeably throughout the Open
edX project. There are many Django microservices in the Open edX ecosystem, such as
- `edx-platform`_ 
- `credentials`_
- `ecommerce`_
- `course-discovery`_

#. If you are creating a new repository, use the `cookiecutter-django-ida`_
   template.

> **TODO:** How to enable translations infrastructure on a Django microservice

Reviewer 2: Is this todo still relevant? Seems like you answer it below.

Author: Yes, it needs expansion and I'll follow up in other PRs.

#. Create a draft pull request in the `openedx-translations repo`_

#. Follow the instructions in the :ref:`testing-in-your-fork` section to verify new repository configuration.

#. Mark your pull request as ready for review and wait for it to be merged.

#. Verify the workflow is syncing the translations properly as described in the :ref:`debugging-translations` section.

#. Run ``OPENEDX_ATLAS_PULL=true make pull_translations`` to verify translations are pulled from the
   `openedx-translations repo`_ into the ``conf/locale`` directory. To generate JavaScript translation files you will

>    likely also need to run ``make static``/``make static.dev``.

Author: Good point @brian-smith-tcril, but I'm reluctant to merge them. 

The three variations would may end up in a large item with 3-bullet points which is slightly more complicated than scrolling up and down.

I'd like to keep it as-is and we can iterate later.

#. If you are creating a new repository, use `frontend-template-application`_.

**TODO:** How to enable translations infrastructure on a React repo

#. Run ``make extract_translations``. Verify that it creates ``src/i18n/transifex_input.json``. This file should be
   excluded from the repo via the ``.gitignore`` file.

#. Create a draft pull request in the `openedx-translations repo`_

#. Follow the instructions in the ref:`testing-in-your-fork` section to verify new repository configuration.

#. Mark your pull request as ready for review and wait for it to be merged.

#. Verify the workflow is syncing the translations properly as described in the :ref:`debugging-translations` section.

#. Depending on how you deploy the micro-frontend, include the ``pull_translations`` make rule with the
   ``OPENEDX_ATLAS_PULL`` environment variable set to ``true`` e.g
   ``$ OPENEDX_ATLAS_PULL=true make pull_translations``.

   This command needs to run before ``npm build`` in order to include updated translations in final micro-frontend
   build.

.. _testing-in-your-fork:

======================================
Testing translations sync in your fork
======================================

> Before submitting a pull request for review in the `openedx-translations repo`_, it's possible to test the workflow
> in your fork by following the steps below:

Reviewer: I think it makes sense to prescriptive here.
```suggestion
Before submitting a pull request for review in the `openedx-translations repo`_, you should test the workflow
on a fork by following the steps below:
```

#. Fork the `openedx-translations repo`_.

> #. Make a pull request in your fork and modify the `extract-translation-source-files.yml`_ workflow to include your

Reviewer: ```suggestion
#. Make a pull request to your fork and modify the `extract-translation-source-files.yml`_ workflow to include your
```

   repo and your organization name ...... **TODO:** add example https://github.com/Zeit-Labs/openedx-translations/pull/1/files

**TODO:** Add full test instructions like Shadi, Brian and Omar tested in their forks.

> #. Add any test translations to your fork of the `openedx-translations repo`_ in the repo directory to overcome the
>    fact that translations don't exist on the upstream `openedx-translations repo`_ yet.

Reviewer: ```suggestion
#. Add any test translations to your fork of the `openedx-translations repo`_ in the repo directory to overcome the
   fact that translations don't exist in the upstream `openedx-translations repo`_ yet.
```

>    We recommend copying an existing translations. For example to test `credentials`_ we would copy the
>    `course discovery translations`_ and modifying it to include `credentials`_ conf/locale.

Reviewer: ```suggestion
   We recommend copying existing translations. For example to test `credentials`_ we would copy the
   `course discovery translations`_ directory and modify it to include `credentials`_ conf/locale.
```

> #. Temporarily edit the ``Makefile`` so the ``pull_translations`` step pulls via your fork e.g.
>    ``atlas pull --repository=Zeit-Labs/openedx-translations``.

Reviewer: ```suggestion
#. Temporarily edit the ``Makefile`` so the ``pull_translations`` step pulls from your fork e.g.
   ``atlas pull --repository=Zeit-Labs/openedx-translations``.
```

#. If you're testing and Open edX plugin, run the ``$ OPENEDX_ATLAS_PULL=true make pull_translations`` command in
   the ``edx-platform`` repo. Otherwise, run ``$ OPENEDX_ATLAS_PULL=true make pull_translations`` in the repository
   you're testing e.g. ``frontend-app-learning``.

#. Run the application (or plugin) and verify the translations you've added are working properly.

>    This step assumes that you're already familiar with `Tutor`_ and/or `devstack`_.

Reviewer: ```suggestion
   **Note:** This step assumes that you're already familiar with `Tutor`_ and/or `devstack`_.
```

.. _debugging-translations:

================================
Debugging translations workflows
================================

> After adding a repository to the `openedx-translations repo`_ verify the following in the next day:

Reviewer: ```suggestion
After adding a repository to the `openedx-translations repo`_ verify the following the next day:
```

#. The `extract-translation-source-files.yml`_ GitHub workflow worked successfully and the build passes in the
   `openedx-translations GitHub Actions tab`_. If something fails, ask for help in the `#wg-translations`_ Open edX
   Slack channel. An example of a successfully generated and merged pull request by the workflow's
   ``edx-transifex-bot`` is the `chore - add updated translation source files #615`_ pull request.

#. Verify that the `openedx-translations project`_ has a new resource for the repo.

#. Ensure the new Transifex resource is 100% translated. Alternatively, Open edX Transifex admins can force sync via
   the "Manual Sync" button in the `Transifex GitHub App sync logs`_ page.

> #. Wait for the next sync to kick in. The sync is managed by Transifex and usually it kicks in less than an hour
>    which we'll verify in the next step. The `Transifex GitHub App sync logs`_ shows the most recent sync results.

Reviewer: ```suggestion
#. Wait for the next sync. The sync is managed by Transifex and usually takes less than an hour
   (which we'll verify in the next step). The `Transifex GitHub App sync logs`_ show the most recent sync results.
```

React Repo
==========

> 1. Create a new repo using the
>    `frontend-template-application <https://github.com/openedx/frontend-template-application>`_ template. If the repo
>    already exists continue to the next steps.

Reviewer: ```suggestion
1. If you are not working within an existing repository, create a new repo using the `frontend-template-application <https://github.com/openedx/frontend-template-application>`_ template.
```
